### ISSUE TITLE
Out of range subscript calling dgemqr from TESTING/LIN/derrtsqr.f

### ISSUE THREAD

**Issue by mickpont** [Reference-LAPACK/lapack#145]
There is a minor problem with the test routine DERRTSQR (file derrtsqr.f)
when it calls subroutine dgemqr(). DERRTSQR declares array TAU to
be of size NMAX (= 2), but DGEMQR examines elements TAU(2) and TAU(3)
 - the array TAU is named T in dgemwqr.f - before it decides whether it
needs to do a quick error return because of an incorrect input argument.

At the moment, because TAU(3) is undefined, in dgemqr.f variable NB
ends up as something random, which might affect the way the rest of the
error checking works.

The test routine DERRTSQR can be fixed easily by increasing the declared
size of array TAU from ( NMAX ) to ( NMAX+1 ), and adding a line
TAU(3) = 1 just after the initialization of TAU(1) and TAU(2).

However, DGEMQR itself oughtn't to examine elements T(2) and T(3) before
it knows that TSIZE is greater than 2. Therefore, dgemgq.f also ought
to be changed a bit. I think that one way to fix it is to replace the lines

      MB = INT( T( 2 ) )
      NB = INT( T( 3 ) )

by something like

      IF (TSIZE.GE.3) THEN
        MB = INT( T( 2 ) )
        NB = INT( T( 3 ) )
      ELSE
        MB = 1
        NB = 1
      END IF

but perhaps there are better ways.

The same issue affects the other precision variants of dgemqr.f
i.e. cgemqr.f, sgemqr.f and zgemqr.f, along with the corresponding
test routines in cerrtsqr.f, serrtsqr.f and zerrtsqr.f.

Mick Pont

**Comment by langou** [comment]
Hi Mike. 

Thanks for the feedback. 

Here is what I propose. 

First there is a problem indeed in DERRTSQR (file derrtsqr.f), we prepare a nice array T( NMAX, NMAX ) and we never use it. I am not sure of the size of T (why is it 2D array? Why NMAX,NMAX?). Then we use TAU instead of T. Mmm. OK so what about removing TAU altogether, and use T, and declare T of size 5. And then, yes we can set NB (so T(3)) to 1. OK. Indeed T(3) was accessed without initialization but this did not have any impact on the code since it was not used. In any case, the routine DERRTSQR (file derrtsqr.f) can be cleaned up. The routine makes `wrong` calls to DGEMQR so as to check that error return works correctly. So typically we call with M < 0 and expect INFO=-1, etc. 

Second about DGEMQR itself. I see your point. So I suggest to have the error checks:
```
      IF( .NOT.LEFT .AND. .NOT.RIGHT ) THEN
        INFO = -1
      ELSE IF( .NOT.TRAN .AND. .NOT.NOTRAN ) THEN
        INFO = -2
      ELSE IF( M.LT.0 ) THEN
        INFO = -3
      ELSE IF( N.LT.0 ) THEN
        INFO = -4
      ELSE IF( K.LT.0 .OR. K.GT.MN ) THEN
        INFO = -5
      ELSE IF( LDA.LT.MAX( 1, MN ) ) THEN
        INFO = -7
      ELSE IF( TSIZE.LT.5 ) THEN
        INFO = -9
      ELSE IF( LDC.LT.MAX( 1, M ) ) THEN
        INFO = -11
```
and an associated 
```
      IF( INFO.NE.0 ) THEN
        CALL XERBLA( 'DGEMQR', -INFO )
        RETURN
      END IF
```
before the infamous lines
```
      MB = INT( T( 2 ) )
      NB = INT( T( 3 ) )
```
And then we would check follow up with computation of workspace, making sure that LWORK is large enough, quick return if workspace query, etc. 

How does this sound? Please let me know. 

If you feel like providing a `pull-request, please feel welcome. Otherwise, I will work on this.

Cheers, 
Julien.

**Comment by mickpont** [comment]
Hi Julien,

All your suggestions look good. I hadn't noticed that there
was an unused array T in DERRTSTQR (presumably copied from
some other test program code). It makes sense to use it now
as a 1D array.

For CGEMQR, Rearranging the way the input arguments are checked in
a way similar to what you suggest is also something we considered, so
please do that. (I guess that the order of the tests in DERRTSTQR
might need rearranging too, to account for the changes in CGEMQR,
but that should be easy).

Thanks for the fast response!

Mick

**Comment by langou** [comment]
In #467, @martin-frbg mentioned that #145 (this issue) was fixed in #427. I am closing this issue now.
